packages/react/src/components/Listbox/ListboxOption.tsx

```diff
@@ -0,0 +1,121 @@
+import React, { forwardRef, useEffect, useCallback } from 'react';
+import classnames from 'classnames';
+import { useId } from 'react-id-generator';
+import { useListboxContext } from './ListboxContext';
+import useSharedRef from '../../utils/useSharedRef';
+
+export type ListboxValue = Readonly<string | number | undefined>;
+
+interface ListboxOptionsProps extends React.HTMLAttributes<HTMLElement> {
+  as?: React.ElementType | string;
+  value?: ListboxValue;
+  disabled?: boolean;
+  activeClass?: string;
+}
+
+function isElementPreceding(a: Element, b: Element) {
+  return !!(b.compareDocumentPosition(a) & Node.DOCUMENT_POSITION_PRECEDING);
+}
+
+const ListboxOption = forwardRef<HTMLElement, ListboxOptionsProps>(
+  (
+    {
+      id: propId,
+      className,
+      as: Component = 'li',
+      children,
+      value,
+      disabled,
+      activeClass = 'ListboxOption--active',
+      onClick,
+      ...props
+    },
+    ref
+  ): JSX.Element => {
+    const { active, selected, setOptions, onSelect } = useListboxContext();
+    const listboxOptionRef = useSharedRef<HTMLElement>(ref);
+    const [id] = propId ? [propId] : useId(1, 'listbox-option');
+    const isActive =
+      active !== null && active.element === listboxOptionRef.current;
+    const isSelected =
+      selected !== null && selected.element === listboxOptionRef.current;
+    const optionValue =
+      typeof value !== 'undefined'
+        ? value
+        : listboxOptionRef.current?.innerText;
+
+    useEffect(() => {
+      const element = listboxOptionRef.current;
+
+      setOptions((options) => {
+        const option = { element, value: optionValue };
+        // istanbul ignore next
+        if (!element) return options;
+
+        // Elements are frequently appended, so check to see if the newly rendered
+        // element follows the last element element first before any other checks
+        if (
+          !options.length ||
+          isElementPreceding(
+            options[options.length - 1].element,
+            option.element
+          )
+        ) {
+          return [...options, option];
+        }
+
+        for (const opt of options) {
+          if (isElementPreceding(element, opt.element)) {
+            const index = options.indexOf(opt);
+            return [
+              ...options.slice(0, index),
+              option,
+              ...options.slice(index)
+            ];
+          }
+        }
+
+        // istanbul ignore next
+        // this should never happen, but just in case fall back to options
+        return options;
+      });
+
+      return () => {
+        setOptions((opts) => opts.filter((opt) => opt.element !== element));
+      };
+    }, [optionValue]);
```

Comment:
Took me some time to understand this block, but I think I get it now. It handles appending `listOptions` at the end of the options or in the middle of it.

Reply:
Yeah, it's a bit tricky. We need to pass the rendered options back up the context but need to sort the list based on dom order here.

I went this route because passing down props via a set of unknown components is gnarly, and this abstracts that away completely.

Reply:
When I saw the context, I also compared it with passing down the props in my mind and I definitely think context is better.